### ISSUE TITLE
[phase-3] SOPD Axis-A5 — rubric-graded self-distill (ROPD machinery, Path-A distillation, never GRPO)

### ISSUE THREAD

**Issue by cklxx** [cklxx/agent-infer#98]
Parent: #90 (SOPD umbrella). **Axis A5 — rubric-graded self-distill.** Path-A (distillation, never GRPO). Gated on #93's `TrajectoryScorer` trait; the open-ended bridge from #93 (exact-match verifiable) to #97 (skills / agentic self-evolving).

**Canonical doc**: [`docs/research/2026-06-14-rubric-opd.md`](https://github.com/cklxx/arle/blob/main/docs/research/2026-06-14-rubric-opd.md) (full grounding). Survey Axis A gains **A5**.

## What & why

A2 (#93) best-of-N self-distill works only where an **exact-match verifier** exists (math/code/tool-schema). Real "skills" (#97 — agentic, tool-use, writing) are **not exact-match-verifiable**. ROPD ([arXiv 2605.07396](https://arxiv.org/abs/2605.07396), May 2026) closes that gap: score on-policy rollouts against an **auto-induced, prompt-specific rubric**, then train on the score. A5 is the rubric generalization of A2's verifier — same best-of-N **distillation** wrapper, rubric Verifier instead of exact-match.

## The hard constraint (load-bearing)

ROPD-as-published optimizes with **GRPO** (rubric score → scalar reward → policy gradient). That collides with the [2026-05-18 OPD-only pivot](https://github.com/cklxx/arle/blob/main/docs/projects/2026-05-18-opd-only-pivot.md) (**never GRPO**). This issue is **Path A only** (confirmed by ckl 2026-06-14: "只用他的策略"): adopt the rubric *machinery* (Rubricator + Verifier + scoring) but drive a **rejection / best-of-N distillation** update — select rubric-best τ*, GKD `λ·CE(student‖τ*) + (1−λ)·KL(student‖EMA)`; losers **dropped** (no negative gradient, no clipped ratio). Default to the **argmax/threshold** form; soft-weight-by-score (reward-weighted regression) is a flagged option, not bring-up. **Literal GRPO-ROPD is declined** — reopening it (= overturning the pivot KILL) is a separate, explicit decision, out of scope for this issue.

## Framework support — code-grounded (what's reuse vs net-new)

| Component | Status | Anchor |
|---|---|---|
| Distillation update (GKD KL+SFT λ-blend; τ* → SFT anchor) | **REUSE** | `opd.rs:1329` `mix_gkd_losses`, `:192` `GkdSftAnchor::StudentRollout` |
| `TrajectoryScorer` trait (rubric impl alongside exact-match) | **NET-NEW** (trait from #93) | #93 defines; #98 adds the rubric scorer |
| Rubricator (induce prompt-specific rubric: contrast EMA-teacher vs student, m≈4 teacher samples) | **NET-NEW** | reuses infer forward + `xgrammar-sys` (grammar-constrained → structured rubric) |
| Verifier (binary per-criterion **blind** scoring) | **NET-NEW** prompt+parse | reuses the judge forward (student / EMA / external) |
| Black-box text-only teacher (no logits) — A4 combo only | **NET-NEW** (external only) | teacher-free EMA has logits, so n/a there; time-share via `offload_engine_weights` |
| Rubric data / prompt source | **NET-NEW** | with the Rubricator, "data" = prompts + a teacher source → connects to #97 |
| Seam / kernel / scheduler | **NONE** | grading is forward-only inference + host selection + the existing backward |

Net: concentrated in the `train` crate + an `xgrammar-sys` dep for structured rubric output. No seam/kernel/scheduler change.

## Two teacher configs (both Path A)

- **Teacher-free (default, extends A1+A2)** — EMA self-teacher is the judge; a **self-Rubricator** induces the rubric by contrasting EMA-teacher answers vs student rollouts (the m≈4 "teacher answers" are EMA samples). No second model.
- **Black-box external teacher (A4)** — the one thing rubric-OPD uniquely unlocks: distill from a **frontier API teacher with no logits** (white-box KL-OPD impossible there). External text-only; time-share device via `offload_engine_weights`.

## Reward-hacking (sharper than exact-match)

Soft rubric judge ⇒ more gameable. ROPD's <2%/self-correcting figure is for a strong external GPT-5.2 judge; a teacher-free EMA judge is weaker. Mitigations: (a) ROPD **blind scoring** (judge can't tell teacher vs student); (b) **EMA-as-judge** (lagged weights harder to game in-step); (c) the #93 tripwire — held-out rubric + held-out open-ended task + manual trajectory spot-check; revert if in-loop score rises while held-out falls.

## Acceptance (license-or-kill)

- **PASS** — held-out lift over the base on **≥2 open-ended dims** (e.g. IFEval + one agentic skill task), multi-seed ≥5, mean±σ + Wilson 95% CI ([2026-05-28 rule](https://github.com/cklxx/arle/blob/main/docs/experience/errors/2026-05-28-mmlu-cross-base-was-noise.md)); reward-hack tripwire holds.
- **KILL** — no CI-separated lift after a sweep, or rubric-gaming can't be contained at acceptable cost ⇒ fall back to A1/A2 (verifiable-only).

## Dependencies

```
#93 (TrajectoryScorer trait + best-of-N wrapper) ──▶ #98 (rubric scorer impl) ──▶ #97 (open-ended skills)
```

Gated on #93 (defines the trait this plugs into). Bridges to #97 (open-ended skills need rubric, not exact-match). Off the Phase-1 batched-lane critical path (#60).

**Comment by cklxx** [comment]
Status update (2026-06-21, verified): rubric machinery **landed** (`crates/train/src/rubric.rs` + `rubric_opd.rs`); 27B rubric-OPD loop **validated end-to-end** (`8c3e0298`, 5 gaps fixed). Remaining: capability curve (`docs/plans/2026-06-21-opd-ceiling-27b-dense.md`) + #93 PASS.

**Comment by cklxx** [comment]
**Re-anchor 2026-08-14.** The body predates the landed rubric-OPD lane: `rubric-opd` exists in `crates/train` and has run end-to-end on H20 with Qwen3.6-27B-FP8 (e.g. the 2026-06-23 train-infer weight-share A/B used the rubric-opd workload). #173's Stage-B item was closed 2026-08-14 as superseded by this pivot. This issue needs a re-scope pass against the current `crates/train` code: mark what of the Rubricator/Verifier/blind-scoring table is now REUSE-landed vs still net-new, and keep only the residual (reward-hack tripwires, A4 black-box external-teacher config). Path-A-only constraint (never GRPO) unchanged.

**Comment by cklxx** [comment]
Host portion landed: c858c0485.

- `Rubricator` (rubric.rs): `induce_prompt(problem, teacher_samples, student_samples)` renders the induction prompt; `parse_induction(output)` parses the JSON rubric, rejecting empty/malformed output (a retry, not a silent default).
- 9 cpu-lane unit tests: verdict parse (accept/reject/process-only/malformed/missing-key), best-of-N selection (accepted/distinct/parse-errors), Rubricator parse (valid/empty/malformed).

The existing primitives (`judge_prompt`, `solve_prompt`, `parse_verdict`, `select`, `select_by_self_consistency`) were already in tree — this tranche adds the missing Rubricator and the test gate. The rubric-OPD orchestration (rubric_opd.rs, cuda-gated) and the multi-seed training gate (held-out lift on ≥2 open-ended dims, reward-hack tripwire) are pod-only. docs/experience/wins/2026-08-25-rubric-primitives.md
